File: Алгоритмы/Kruskal/src/Kruskal.java
import java.util.Arrays;
import java.util.Scanner;

public class Kruskal {

    static class Vertex implements Comparable<Vertex>{
        int x, y;
        int dep;
        Vertex parent;

        public Vertex(int x, int y) {
            this.x = x;
            this.y = y;
            dep = 0;
            parent = this;
        }

        public int compareTo(Vertex obj){
            if (this.x > obj.x && this.y > obj.y)
                return 1;
            if (this.x == obj.x && this.y == obj.y)
                return 0;
            else return -1;
        }

        public Vertex find(){
            if(compareTo(this.parent) == 0)
                return this;
            else
                return this.parent.find();
        }
    }

    static class Road implements Comparable<Road>{
        Vertex A;
        Vertex B;
        double len;

        public Road(Vertex A, Vertex B) {
            this.A = A;
            this.B = B;
            //len = 0;
        }

        @Override
        public int compareTo(Road obj) {
            if (this.len > obj.len)
                return 1;
            if (this.len == obj.len)
                return 0;
            else
                return -1;
        }
    }

    public static void union(Vertex v, Vertex u){
        Vertex v1 = v.find();
        Vertex u1 = u.find();
        if(v1.dep > u1.dep)
            u1.parent = v1;
        else{
            v1.parent = u1;
            u1.dep++;
        }
    }

    public static void main(String[] args) {
        long start = System.currentTimeMillis();
        Scanner in = new Scanner(System.in);
        int n = in.nextInt();
        Vertex[] nodes = new Vertex[n];
        for (int i = 0; i < n; i++) {
            int x = in.nextInt();
            int y = in.nextInt();
            nodes[i] = new Vertex(x, y);
        }

        int num = n * (n - 1) / 2;
        Road[] roads = new Road[num];
        int k = 0;
        for (int i = 0; i < n; i++)
            for (int j = i + 1; j < n; j++) {
                roads[k] = new Road(nodes[i], nodes[j]);
                roads[k].len = Math.sqrt(Math.pow(nodes[i].x - nodes[j].x, 2) + Math.pow(nodes[i].y - nodes[j].y, 2));
                k++;
            }
        Arrays.sort(roads);
            //for(int i = 0; i < num; i++)
                //System.out.println(roads[i].len);
         double summ = 0;
         int edges = 0;
         for(int i = 0; i < num && edges < n - 1; i++) {
             //System.out.println(roads[i].A.find().x + " " + roads[i].A.find().y);
             //System.out.println(roads[i].B.find().x + " " + roads[i].B.find().y);
             //System.out.println();
             if (roads[i].A.find().compareTo(roads[i].B.find()) != 0) {
                 summ += roads[i].len;
                 union(roads[i].A, roads[i].B);
                 edges++;
             }
         }
          System.out.printf("%.2f\n", summ);
         System.out.println((double)(System.currentTimeMillis() - start));
    }
}

File: Алгоритмы/Kruskal.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

struct Vertex{
        int x, y;
        int dep;
        struct Vertex *parent;
    };

int compare(struct Vertex *this, struct Vertex *obj){
	if (this->x > obj->x && this->y > obj->y)
                return 1;
        if (this->x == obj->x && this->y == obj->y)
                return 0;
        else return -1;
        }
        
struct Vertex* find(struct Vertex* this){
        if(compare(this, this->parent) == 0)
                return this;
        else
                return find(this->parent);
}
        
struct Road{
        struct Vertex *A;
        struct Vertex *B;
        double len;
};
    
struct Vertex *nodes;
struct Road *roads;

int compareTo(const struct Road *this, const struct Road *obj) {
            if (this->len > obj->len)
                return 1;
            if (this->len == obj->len)
                return 0;
            else
                return -1;
}

void Union(struct Vertex *v, struct Vertex *u){
        struct Vertex *v1 = find(v);
        struct Vertex *u1 = find(u);
        if(v1->dep > u1->dep)
            u1->parent = v1;
        else{
            v1->parent = u1;
            u1->dep++;
        }
    }

int main() {
	int n;
        scanf("%d", &n);
        nodes = (struct Vertex*)malloc(sizeof(struct Vertex)*n);
        for (int i = 0; i < n; i++) {
            int x, y;
            scanf("%d %d", &x, &y);
            nodes[i].x = x;
            nodes[i].y = y;
            nodes[i].parent = &nodes[i];
            nodes[i].dep = 0;
        }

        int num = n * (n - 1) / 2;
        roads = (struct Road*)malloc(sizeof(struct Road)*num);
        int k = 0;
        for (int i = 0; i < n; i++)
            for (int j = i + 1; j < n; j++) {
                roads[k].A = &nodes[i];
                roads[k].B = &nodes[j];
                int a = pow((roads[k].A->x - roads[k].B->x), 2);
                int b = pow((roads[k].A->y - roads[k].B->y), 2);
                roads[k].len = sqrt(a + b);
                k++;
            }
        qsort(roads,num, sizeof(struct Road), (int(*) (const void *, const void *)) compareTo);
         double summ = 0;
         int edges = 0;
         for(int i = 0; i < num && edges < n - 1; i++) {
             if (compare(find(roads[i].A), find(roads[i].B)) != 0) {
                 summ += roads[i].len;
                 Union(roads[i].A, roads[i].B);
                 edges++;
             }
         }
          printf("%.2f\n", summ);
    }

